Fixed Forecast and Flower Gift

AgustinGDLV opened the pull request:
## Description
This PR does the following:
- The now unused `u8 TryWeatherFormChange` has been removed (last remnant of old Castform hacks?).
- Flower Gift now correctly boosts its own and an ally's attack and special defense by 1.5x.
- Forecast and Flower Gift form changes work consistently with weather changes.
- Gastro Acid, Core Enforcer, Simple Beam, and Worry Seed revert Castform and Cherrim to normal in Gen 5+ (config added).
- Adds tests for Forecast and Flower Gift.

## Tests Added
![image](https://user-images.githubusercontent.com/103095241/236355112-335131c2-7b26-455c-984d-5c2752545a6b.png)
![image](https://user-images.githubusercontent.com/103095241/236355126-de8189cf-ceca-48b7-97c0-8d34db12c130.png)

## Issue(s) that this PR fixes
Fixes #2972.

## **Discord contact info**
Agustin#1522

AsparagusEduardo commented on data/battle_scripts_1.s:
```diff
@@ -8663,14 +8685,15 @@ BattleScript_ShedSkinActivates::
+	savetarget
+	setbyte gBattlerTarget, 0
 BattleScript_ActivateWeatherAbilities_Loop:
-	activateweatherchangeabilities BS_ATTACKER
+	copybyte sBATTLER, gBattlerTarget
+	activateweatherchangeabilities BS_TARGET
 BattleScript_ActivateWeatherAbilities_Increment:
-	addbyte gBattlerAttacker, 1
-	jumpifbytenotequal gBattlerAttacker, gBattlersCount, BattleScript_ActivateWeatherAbilities_Loop
-	copybyte gBattlerAttacker, sBATTLER
+	addbyte gBattlerTarget, 1
+	jumpifbytenotequal gBattlerTarget, gBattlersCount, BattleScript_ActivateWeatherAbilities_Loop
+	restoretarget
```
This can potentially break other abilities that trigger on weather change (Protosynthesis and Ice Face).
Ice face has its own check to prevent its trigger on more than 1 weather activation, but Protosynthesis doesn't, so it might trigger on each Sun turn

AgustinGDLV replied:
Protosynthesis has a check to `gSpecialStatuses[battlerId].weatherAbilityDone` that isn't reset from turn to turn and is only used for Protosynthesis.

AsparagusEduardo commented on test/ability_flower_gift.c:
```diff
@@ -0,0 +1,86 @@
+
+SINGLE_BATTLE_TEST("Flower Gift transforms Cherrim back to normal when its ability is suppressed")
+{
+    GIVEN {
```
Needs an assumption for `B_WEATHER_FORM_SUPPRESS >= GEN_5`.

AsparagusEduardo replied:
Still needs the assumption (though for `B_WEATHER_FORM` instead now), otherwise it'll cause this to happen when switched:
```
Flower Gift transforms Cherrim back to normal when its ability is suppressed: FAIL
test/ability_flower_gift.c:48: Unmatched ABILITY_POPUP
Flower Gift transforms Cherrim in harsh sunlight: FAIL
test/ability_flower_gift.c:12: Unmatched ABILITY_POPUP
Flower Gift increases the Sp. Def of Cherrim and its allies by 1.5x 2/2: FAIL
test/ability_flower_gift.c:108: Unmatched ABILITY_POPUP
Flower Gift increases the attack of Cherrim and its allies by 1.5x 2/2: FAIL
test/ability_flower_gift.c:75: Unmatched ABILITY_POPUP
Flower Gift transforms Cherrim back to normal when weather changes: FAIL
test/ability_flower_gift.c:28: Unmatched ABILITY_POPUP
Forecast transforms Castform back to normal when its ability is suppressed: FAIL
test/ability_forecast.c:251: Unmatched ANIMATION
```

AsparagusEduardo commented on test/ability_flower_gift.c:
```diff
@@ -0,0 +1,86 @@
+    }
+}
+
+DOUBLE_BATTLE_TEST("Flower Gift increases the attack of Cherrim's ally by 1.5x", s16 damage)
```
> During [harsh sunlight](https://bulbapedia.bulbagarden.net/wiki/Harsh_sunlight), Flower Gift increases the [Attack](https://bulbapedia.bulbagarden.net/wiki/Stat#Attack) and [Special Defense](https://bulbapedia.bulbagarden.net/wiki/Stat#Special_Defense) stats of the Pokémon with this Ability and its allies by 1.5×.

Bulbapedia's description imply the necessity of 3 more tests, one for its own Attack, its own Special Defense and the Ally's Special Defense.

AsparagusEduardo commented on test/ability_forecast.c:
```diff
@@ -0,0 +1,201 @@
+#include "global.h"
+#include "test_battle.h"
```
Needs a test for overriting Forecast like how Flower Gift does.

AsparagusEduardo commented on test/ability_forecast.c:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+SINGLE_BATTLE_TEST("Forecast transforms Castform in weather from an ability")
```
I'd parametrize with all 3 relevant weather abilities.

AsparagusEduardo commented on test/ability_forecast.c:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+SINGLE_BATTLE_TEST("Forecast transforms Castform in primal weather")
```
I'd parametrize with both relevant weather abilities.

AsparagusEduardo commented on test/ability_forecast.c:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+SINGLE_BATTLE_TEST("Forecast transforms Castform back to normal when Sandstorm is active")
```
I'd add a test for Strong Winds as well.

AgustinGDLV commented on src/battle_script_commands.c:
```diff
@@ -16559,3 +16563,18 @@ void BS_ItemRestorePP(void) {
+    u16 ability = gBattleMons[gBattlerTarget].ability;
+    if (B_WEATHER_FORM_SUPPRESS >= GEN_5
+        && (ability == ABILITY_FORECAST || ability == ABILITY_FLOWER_GIFT)
+        && TryBattleFormChange(gBattlerTarget, FORM_CHANGE_BATTLE_SWITCH)) // revert form
```
What do you recommend I use to turn something into its base form? There’s a small chance I do something janky like this in Dynamax that’ll I’ll have to check and fix 👍

AsparagusEduardo replied:
I think the way to go would be to have ability as optional parameter for `FORM_CHANGE_BATTLE_WEATHER` in the form change tables. If it has it defined, it will require the ability to maintain the form.

AgustinGDLV commented on src/battle_script_commands.c:
```diff
@@ -16559,3 +16563,18 @@ void BS_ItemRestorePP(void) {
+{
+    NATIVE_ARGS();
+    u16 ability = gBattleMons[gBattlerTarget].ability;
+    if (B_WEATHER_FORM_SUPPRESS >= GEN_5
```
`B_WEATHER_FORM_SUPPRESS` is written as a generation config and so I don't think I can just write `#ifdef B_WEATHER_FORM_SUPPRESS` since it's always defined right now.

AgustinGDLV commented:
Ready for review! I added a missing interaction with Air Lock (that also makes it pretty much impossible to test Strong Winds), reworked how the weather form reversion is done, and added all the missing tests. Castform is now functional with snow, too!

AsparagusEduardo commented on src/battle_util.c:
```diff
@@ -10246,9 +10158,25 @@ u16 GetBattleFormChangeTargetSpecies(u8 battlerId, u16 method)
+                        targetSpecies = formChanges[i].targetSpecies;
+                    }
+                    // We need to revert the weather form if the field is under Air Lock, too.
+                    if (!WEATHER_HAS_EFFECT && formChanges[i].param1 == B_WEATHER_NONE)
```
Shouldn't this be an `else if`?

AgustinGDLV replied:
Done, ready for review.

AsparagusEduardo reviewed (changes requested):
`B_WEATHER_FORMS` already exists.

In retrospect, I didn't properly communicate that I wanted Castform and Cherrim's generational changes to be handled by the form change tables, which is why I proposed the optional ability parameter, in order to toggle between them by changing the data.

AsparagusEduardo commented on src/data/pokemon/form_change_tables.h:
```diff
@@ -156,11 +156,19 @@ static const struct FormChange sAbsolFormChangeTable[] = {
 };
 
 static const struct FormChange sCastformFormChangeTable[] = {
+#ifdef B_WEATHER_FORMS >= GEN_5
```
```suggestion
#if B_WEATHER_FORMS >= GEN_5
```

AgustinGDLV commented:
Added assumptions, ready again.
